clippy_lints/src/format_args.rs

```diff
@@ -86,11 +147,55 @@ impl<'tcx> LateLintPass<'tcx> for FormatArgs {
                     check_format_in_format_args(cx, outermost_expn_data.call_site, name, arg.param.value);
                     check_to_string_in_format_args(cx, name, arg.param.value);
                 }
+
+                if !self.format_args_capture {
+                    return;
+                }
+                let mut inline_spans = Vec::new();
+                // If any of the arguments are referenced by an index number,
+                // and that argument is not a simple variable and cannot be inlined,
+                // we cannot remove any other arguments in the format string,
+                // because the index numbers might be wrong after inlining.
+                // Example of an un-inlinable format:  print!("{}{1}", foo, 2)
+                let do_inlining = format_args.params().all(|p| check_inline(cx, &p, &mut inline_spans));
+                if do_inlining && !inline_spans.is_empty()
+                {
+                    span_lint_and_then(
+                        cx,
+                        NEEDLESS_FORMAT_ARGS,
+                        outermost_expn_data.call_site,
+                        "variables can be used directly in the `format!` string",
+                        |diag| {
+                            diag.multipart_suggestion("change this to", inline_spans, Applicability::MachineApplicable);
+                        },
+                    );
+                }
             }
         }
     }
```

Comment:
```suggestion
    }
    
    extract_msrv_attr!(LateContext);
```

Reply:
Aha!  This is the piece i was missing. Writing code with macros... oh well. Thanks!!!